contracts/src/v0.8/dev/KeeperRegistryBase.sol

```diff
@@ -191,11 +215,24 @@ abstract contract KeeperRegistryBase is ConfirmedOwner, ExecutionPrevention, Ree
   function _calculatePaymentAmount(
     uint256 gasLimit,
     uint256 gasWei,
-    uint256 linkEth
+    uint256 linkEth,
+    bool isExecution,
+    bytes memory txCallData
   ) internal view returns (uint96 payment) {
     uint256 weiForGas = gasWei * (gasLimit + REGISTRY_GAS_OVERHEAD);
     uint256 premium = PPB_BASE + s_storage.paymentPremiumPPB;
-    uint256 total = ((weiForGas * (1e9) * (premium)) / (linkEth)) + (uint256(s_storage.flatFeeMicroLink) * (1e12));
+    uint256 l1CostWei = 0;
+    if (PAYMENT_MODEL == PaymentModel.OPTIMISM) {
+      l1CostWei = OPTIMISM_ORACLE.getL1Fee(txCallData);
+    } else if (PAYMENT_MODEL == PaymentModel.ARBITRUM) {
+      l1CostWei = ARB_NITRO_ORACLE.getCurrentTxL1GasFees();
```

Comment:
imho, this `PaymentModel` is just another setting that can be mis-configured and I think we can better better protect against misuse - and keep our code cleaner - by making the `KeeperRegistry` inheritable. Currently 4 different functions have to be augmented for compatibility: `_calculatePaymentAmount`, `_performUpkeepWithParams`, `getMaxPaymentForGas`, and `_generatePerformParams`. I think if we make the contracts inheritable, each contract only has to define it's own `_calculatePaymentAmount` method.

Reply:
Thanks for raising this @RyanRHall !

I think these are two separate concerns
- Whether we should have separate contracts for OP and Arb which inherit from some base contract vs having an IF statement. I think there are tradeoffs for both, separate contracts also add an overhead in deployment and maintaining separate versions (arbRegistry1_1 ...) + add room for mistakes, e.g. deploying a wrong registry on a wrong network. Arguably this is equivalent to the overhead of another constructor param. Anyway this is a minor issue in my mind and something we can evaluate later.

- Whether we can contain all changes within `_calculatePaymentAmount`. This I think we should be able to do irrespective of above. @FelixFan1992 wdyt about removing `txCallData` param in `_calculatePaymentAmount` and generating it inside the function? Would limit the surface area of changes to other functions

Reply:
Another thought: There are two places where we have different logic dependent upon `isExecution`. `_adjustGasPrice` and `_calculatePaymentAmount`.

If we can subsume the logic of `_adjustGasPrice` within `_calculatePaymentAmount` it would simplify the logic even further for callsites